<div>
<img src=https://www.institutedata.com/wp-content/uploads/2019/10/iod_h_tp_primary_c.svg width="300">
</div>

## Lab 4.2.1: Feature Selection I

### 1. Load & Explore Data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

#### 1.1 Load Data

In [2]:
# Read CSV
wine_csv = 'dat/winequality_merged.csv'
original_data = pd.read_csv(wine_csv)

#### 1.2 Explore Data (Exploratory Data Analysis)

In [3]:
# ANSWER
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         6497 non-null   float64
 1   volatile acidity      6497 non-null   float64
 2   citric acid           6497 non-null   float64
 3   residual sugar        6497 non-null   float64
 4   chlorides             6497 non-null   float64
 5   free sulfur dioxide   6497 non-null   float64
 6   total sulfur dioxide  6497 non-null   float64
 7   density               6497 non-null   float64
 8   pH                    6497 non-null   float64
 9   sulphates             6497 non-null   float64
 10  alcohol               6497 non-null   float64
 11  quality               6497 non-null   int64  
 12  red_wine              6497 non-null   int64  
dtypes: float64(11), int64(2)
memory usage: 660.0 KB


In [4]:
original_data

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
4               7.4              0.70         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
6492            6.2              0.21         0.29             1.6      0.039   
6493            6.6              0.32         0.36             8.0      0.047   
6494            6.5              0.24         0.19             1.2      0.041   
6495            5.5              0.29         0.30             1.1      0.022   
6496            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
6492                 24.0                  92.0  0.99114  3.27       0.50   
6493                 57.0                 168.0  0.99490  3.15       0.46   
6494                 30.0                 111.0  0.99254  2.99       0.46   
6495                 20.0                 110.0  0.98869  3.34       0.38   
6496                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  red_wine  
0         9.4        5         1  
1         9.8        5         1  
2         9.8        5         1  
3         9.8        6         1  
4         9.4        5         1  
...       ...      ...       ...  
6492     11.2        6         0  
6493      9.6        5         0  
6494      9.4        6         0  
6495     12.8        7         0  
6496     11.8        6         0  

[6497 rows x 13 columns]

In [5]:
original_data.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
red_wine                0
dtype: int64

In [6]:
original_data.duplicated().sum()

1177

In [7]:
original_data[original_data.duplicated()]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
4               7.4             0.700         0.00            1.90      0.076   
11              7.5             0.500         0.36            6.10      0.071   
27              7.9             0.430         0.21            1.60      0.106   
40              7.3             0.450         0.36            5.90      0.074   
65              7.2             0.725         0.05            4.65      0.086   
...             ...               ...          ...             ...        ...   
6427            6.4             0.230         0.35           10.30      0.042   
6449            7.0             0.360         0.35            2.50      0.048   
6450            6.4             0.330         0.44            8.90      0.055   
6455            7.1             0.230         0.39           13.70      0.058   
6479            6.6             0.340         0.40            8.10      0.046   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
4                    11.0                  34.0  0.99780  3.51       0.56   
11                   17.0                 102.0  0.99780  3.35       0.80   
27                   10.0                  37.0  0.99660  3.17       0.91   
40                   12.0                  87.0  0.99780  3.33       0.83   
65                    4.0                  11.0  0.99620  3.41       0.39   
...                   ...                   ...      ...   ...        ...   
6427                 54.0                 140.0  0.99670  3.23       0.47   
6449                 67.0                 161.0  0.99146  3.05       0.56   
6450                 52.0                 164.0  0.99488  3.10       0.48   
6455                 26.0                 172.0  0.99755  2.90       0.46   
6479                 68.0                 170.0  0.99494  3.15       0.50   

        alcohol  quality  red_wine  
4      9.400000        5         1  
11    10.500000        5         1  
27     9.500000        5         1  
40    10.500000        5         1  
65    10.900000        5         1  
...         ...      ...       ...  
6427   9.200000        5         0  
6449  11.100000        6         0  
6450   9.600000        5         0  
6455   9.000000        6         0  
6479   9.533333        6         0  

[1177 rows x 13 columns]

In [8]:
duplicate_index = original_data[original_data.duplicated()].index

In [9]:
nonduplicate_data = original_data.drop(duplicate_index)
nonduplicate_data                                         

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
5               7.4              0.66         0.00             1.8      0.075   
...             ...               ...          ...             ...        ...   
6492            6.2              0.21         0.29             1.6      0.039   
6493            6.6              0.32         0.36             8.0      0.047   
6494            6.5              0.24         0.19             1.2      0.041   
6495            5.5              0.29         0.30             1.1      0.022   
6496            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
5                    13.0                  40.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
6492                 24.0                  92.0  0.99114  3.27       0.50   
6493                 57.0                 168.0  0.99490  3.15       0.46   
6494                 30.0                 111.0  0.99254  2.99       0.46   
6495                 20.0                 110.0  0.98869  3.34       0.38   
6496                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  red_wine  
0         9.4        5         1  
1         9.8        5         1  
2         9.8        5         1  
3         9.8        6         1  
5         9.4        5         1  
...       ...      ...       ...  
6492     11.2        6         0  
6493      9.6        5         0  
6494      9.4        6         0  
6495     12.8        7         0  
6496     11.8        6         0  

[5320 rows x 13 columns]

In [ ]:
# Copied code from seaborn examples
# https://seaborn.pydata.org/examples/many_pairwise_correlations.html
sns.set(style="white")

# Generate a mask for the upper triangle
mask = np.zeros_like(wine_corr, dtype=np.bool)
mask[np.triu_indices_from(mask)] = True

# Set up the matplotlib figure
f, ax = plt.subplots(figsize=(15, 10))

# Generate a custom diverging colormap
cmap = sns.diverging_palette(220, 10, as_cmap=True)

# Draw the heatmap with the mask and correct aspect ratio
sns.heatmap(wine_corr, mask=mask, cmap=cmap, vmax=.3, center=0,
            square=True, linewidths=.5, cbar_kws={"shrink": .5}, annot=True)

plt.show()

### 2. Set Target Variable

Create a target variable for wine quality.

In [11]:
# Target Variable
y = original_data['quality']
y

0       5
1       5
2       5
3       6
4       5
       ..
6492    6
6493    5
6494    6
6495    7
6496    6
Name: quality, Length: 6497, dtype: int64

### 3. Set Predictor Variables

Create a predictor matrix with variables of your choice. State your reason.

In [12]:
# ANSWER
#X = original_data.drop(['alcohol', 'residual sugar', 'fixed acidity', 'volatile acidity', 'citric acid'],axis=1).values
X = original_data.drop(['chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'quality', 'red_wine', 'fixed acidity'],axis=1)
X


volatile acidity  citric acid  residual sugar  alcohol
0                 0.70         0.00             1.9      9.4
1                 0.88         0.00             2.6      9.8
2                 0.76         0.04             2.3      9.8
3                 0.28         0.56             1.9      9.8
4                 0.70         0.00             1.9      9.4
...                ...          ...             ...      ...
6492              0.21         0.29             1.6     11.2
6493              0.32         0.36             8.0      9.6
6494              0.24         0.19             1.2      9.4
6495              0.29         0.30             1.1     12.8
6496              0.21         0.38             0.8     11.8

[6497 rows x 4 columns]

### 4. Using Linear Regression Create a Model and Test Score

In [13]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

In [14]:
# Train-Test Split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [15]:
# Create a model for Linear Regression
reg = LinearRegression()

# Fit the model with the Training data
reg.fit(X_train, y_train)

# Make predictions using the testing set
y_pred_test = reg.predict(X_test)
y_pred_train = reg.predict(X_train)

# Calculate the score (R^2 for Regression) for Training Data
print('Coefficient of determination: %f \n'
      % r2_score(y_train, y_pred_train))

# Calculate the score (R^2 for Regression) for Testing Data
print('Coefficient of determination: %f \n'
      % r2_score(y_test, y_pred_test))

# The coefficients
print('Coefficients: \n', reg.coef_)
# The mean squared error
print('\nMean squared error: %f'
      % mean_squared_error(y_test, y_pred_test))

print(y_pred_test)
print(y_pred_train)

Coefficient of determination: 0.268125 

Coefficient of determination: 0.254345 

Coefficients: 
 [-1.25263359 -0.05397053  0.01625682  0.34722493]

Mean squared error: 0.550701
[6.71643215 5.18535918 6.30520919 ... 6.64062702 5.43765623 5.58293079]
[5.97660418 6.48817186 5.76030744 ... 5.44049315 5.60110804 5.08774586]


### BONUS: Cross validation

In [16]:
# Cross validation 
from sklearn.model_selection import KFold 
from sklearn.metrics import mean_squared_error

In [23]:
# Set up 5-fold cross validation  
k_fold = KFold(5, shuffle=True)
train_scores = []
train_rmse = []
test_scores = []
test_rmse = []

for k, (train, test) in enumerate(k_fold.split(X)):
    
    # Get training and test sets for X and y
    X_train = X.iloc[train, ]
    y_train = y.iloc[train, ]
    X_test = X.iloc[test, ]
    y_test = y.iloc[test, ]
    
    # Fit model with training set
    reg.fit(X_train, y_train)
    
    # Make predictions with training and test set
    y_pred_test = reg.predict(X_test)
    y_pred_train = reg.predict(X_train)
    
    # Score R2 and RMSE on training and test sets and store in list
    
    # Calculate the score (R^2 for Regression) for Training Data
    train_scores.append(r2_score(y_train, y_pred_train))

    # Calculate the score (R^2 for Regression) for Testing Data
    test_scores.append(r2_score(y_test, y_pred_test))
    
    # The mean squared error - test
    test_rmse.append(mean_squared_error(y_test, y_pred_test))
    
    # The mean squared error - train
    train_rmse.append(mean_squared_error(y_train, y_pred_train))
    
    
# Create a metrics_df dataframe to display r2 and rmse scores
metrics_df = pd.DataFrame(list(zip(train_scores, test_scores, test_rmse, train_rmse)),columns =['train_scores', 'test_scores','test_rmse','train_rmse'])
metrics_df

train_scores  test_scores  test_rmse  train_rmse
0      0.265606     0.265213   0.559254    0.560147
1      0.266367     0.260755   0.569247    0.557732
2      0.267471     0.257557   0.572088    0.556991
3      0.264712     0.268989   0.529351    0.567647
4      0.264041     0.270872   0.571930    0.556986

In [24]:
# Use .Describe on the metrics
metrics_df.describe()

train_scores  test_scores  test_rmse  train_rmse
count      5.000000     5.000000   5.000000    5.000000
mean       0.265639     0.264677   0.560374    0.559901
std        0.001351     0.005553   0.018119    0.004520
min        0.264041     0.257557   0.529351    0.556986
25%        0.264712     0.260755   0.559254    0.556991
50%        0.265606     0.265213   0.569247    0.557732
75%        0.266367     0.268989   0.571930    0.560147
max        0.267471     0.270872   0.572088    0.567647

### 5. Feature Selection

What's your score (R^2 for Regression) for Testing Data? 

How many feature have you selected? Can you improve your score by selecting different features?

Coefficient of determination: 0.254345

In [28]:
X

volatile acidity  citric acid  residual sugar  alcohol
0                 0.70         0.00             1.9      9.4
1                 0.88         0.00             2.6      9.8
2                 0.76         0.04             2.3      9.8
3                 0.28         0.56             1.9      9.8
4                 0.70         0.00             1.9      9.4
...                ...          ...             ...      ...
6492              0.21         0.29             1.6     11.2
6493              0.32         0.36             8.0      9.6
6494              0.24         0.19             1.2      9.4
6495              0.29         0.30             1.1     12.8
6496              0.21         0.38             0.8     11.8

[6497 rows x 4 columns]

**Please continue with Lab 4.2.2 with the same dataset.**



---



---



> > > > > > > > > © 2021 Institute of Data


---



---



